### PR TITLE
Fix: Images - Thumbnail files not deleted upon deleting its main image/attachment [ED-6973]

### PR DESCRIPTION
**Pull request by nahmana**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -29,5 +29,10 @@
 		{
 			"name": "Elementor team"
 		}
-	]
+	],
+	"config": {
```
Comment (StyleShit) [inline composer.json:33]: I think it needs to be moved to another PR @matipojo

File: core/utils/collection.php [hunk 2]
```diff
@@ -118,6 +118,12 @@ public function map( callable $callback ) {
 		return new static( array_combine( $keys, $items ) );
 	}
 
+	public function each( callable $callback ) {
```
Comment (StyleShit) [inline core/utils/collection.php:121]: Maybe add a PHPDoc, just for consistency with other functions

File: includes/managers/image.php [hunk 3]
```diff
@@ -82,7 +84,23 @@ public function get_details( $id, $size, $is_first_time ) {
...
+				$image_meta = wp_get_attachment_metadata( $id );
+
+				$image_meta['sizes'][ 'elementor_' . $size ] = [
+					'file' => $thumbs_path,
+					'width' => $matches[1],
+					'height' => $matches[2],
+					'mime-type' => get_post_mime_type( $id ),
+				];
+
+				// Attach custom image to original.
```
Comment (StyleShit) [inline includes/managers/image.php:100]: Redundant comment. If you really want, put that above the array assignment (Line 93)

File: includes/managers/image.php [hunk 4]
```diff
@@ -130,6 +148,21 @@ public function get_lightbox_image_attributes( $id ) {
 		return $attributes;
 	}
 
+	private function delete_custom_images( $post_id ) {
+		$image_meta = wp_get_attachment_metadata( $post_id );
+
+		( new Collection( $image_meta['sizes'] ) )
+			->filter( function ( $value, $key ) {
+				return ( 0 === strpos( $key, 'elementor_custom_' ) );
```
Comment (StyleShit) [inline includes/managers/image.php:156]: Maybe you should use the same regex as above, just to be strict? IDK

File: tests/phpunit/elementor/core/utils/test-collection.php [hunk 5]
```diff
@@ -410,4 +410,25 @@ public function test_sort_keys() {
...
+		$mock->expects( $this->exactly( 3 ) )
+			->method( 'each_callback' )
+			->withConsecutive(
+				[ 1, 'a' ],
+				[ 2, 'b' ],
+				[ 3, 'c' ]
+			);
+
+		// Act
+		$collection->each( [$mock, 'each_callback'] );
```
Comment (StyleShit) [inline tests/phpunit/elementor/core/utils/test-collection.php:432]:
```suggestion
		$collection->each( [ $mock, 'each_callback' ] );
```

File: tests/phpunit/elementor/includes/managers/test-image.php [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+		$attachment_id = $this->_make_attachment( [
+			'file' => __DIR__ . '/mock/mock-image.png',
+			'url' => 'https://test.local/image.png',
+		] );
+
+		$image_meta = wp_get_attachment_metadata( $attachment_id );
+
+		$image_meta['sizes']['test'] = [
+			'file' => 'base-test-image.png',
+			'width' => 300,
```
Comment (StyleShit) [inline tests/phpunit/elementor/includes/managers/test-image.php:23]:
```suggestion
		$image_meta['sizes']['test_size'] = [
			'file' => 'test-image.png',
			'width' => 300,
```

IDK

File: tests/phpunit/elementor/includes/managers/test-image.php [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+		$new_metadata = wp_get_attachment_metadata( $attachment_id );
+
+		$this->assertTrue( is_array( $new_metadata['sizes'] ) );
+		$this->assertEquals( [
+				'file' => 'base-test-image.png',
+				'width' => '300',
+				'height' => '300',
+				'mime-type' => 'image/png',
+		], $new_metadata['sizes']['test'] );
+		$this->assertEquals( [
```
Comment (StyleShit) [inline tests/phpunit/elementor/includes/managers/test-image.php:50]: add some space

File: tests/phpunit/elementor/includes/managers/test-image.php [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+			'width' => 100,
+			'height' => 100,
+			'mime-type' => 'image/png',
+		];
+
+		// Attach custom image to original.
+		wp_update_attachment_metadata( $attachment_id, $image_meta );
+
+		// Register
+		$images_manager = new Images_Manager();
```
Comment (StyleShit) [inline tests/phpunit/elementor/includes/managers/test-image.php:85]: redundant comments

File: tests/phpunit/elementor/includes/managers/test-image.php [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+		// Act
+		do_action( 'delete_attachment', $attachment_id );
+
+		// Assert.
+		$this->assertTrue( file_exists( $base_full_path ) );
+		$this->assertFalse( file_exists( $custom_full_path ) );
```
Comment (StyleShit) [inline tests/phpunit/elementor/includes/managers/test-image.php:92]:
```suggestion
		$this->assertFileExists( $base_full_path );
		$this->assertFileDoesNotExist( $custom_full_path );
```

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: includes/managers/image.php (hunk 3 above)
Comment (nahmana) [reply includes/managers/image.php:100]: Done,

File: core/utils/collection.php (hunk 2 above)
Comment (nahmana) [reply core/utils/collection.php:121]: Done.

File: tests/phpunit/elementor/core/utils/test-collection.php (hunk 5 above)
Comment (nahmana) [reply tests/phpunit/elementor/core/utils/test-collection.php:432]: Done.

File: tests/phpunit/elementor/includes/managers/test-image.php (hunk 7 above)
Comment (nahmana) [reply tests/phpunit/elementor/includes/managers/test-image.php:50]: Done.

File: tests/phpunit/elementor/includes/managers/test-image.php (hunk 8 above)
Comment (nahmana) [reply tests/phpunit/elementor/includes/managers/test-image.php:85]: Done.

File: tests/phpunit/elementor/includes/managers/test-image.php (hunk 9 above)
Comment (nahmana) [reply tests/phpunit/elementor/includes/managers/test-image.php:92]: Done.

File: tests/phpunit/elementor/includes/managers/test-image.php (hunk 6 above)
Comment (nahmana) [reply tests/phpunit/elementor/includes/managers/test-image.php:23]: Done.

**Comment by StyleShit** [comment]
Your tests are failing in some PHP versions due to the change of the "file_exists" assertion I told you to change. 
So just revert that 🤷

File: composer.json (hunk 1 above)
Comment (nahmana) [reply composer.json]: Done.

File: includes/managers/image.php (hunk 4 above)
Comment (nahmana) [reply includes/managers/image.php:156]: No, the regex is used there for retrieving the dimensions into variables etc...

File: core/utils/collection.php [hunk 10]
```diff
@@ -118,6 +118,18 @@ public function map( callable $callback ) {
...
+	/**
+	 * Run a callback over each of the items.
+	 *
+	 * @param callable $callback
+	 * @return void
+	 */
+	public function each( callable $callback ) {
+		foreach ( $this->items as $key => $value ) {
+			$callback( $value, $key );
+		}
```
Comment (matipojo) [inline core/utils/collection.php:130]: Please allow chaining like `$collection->each(...)->map(...)`
```suggestion
		}
	return new static( $result );
```

File: tests/phpunit/elementor/core/utils/test-collection.php [hunk 11]
```diff
@@ -410,4 +410,25 @@ public function test_sort_keys() {
...
+		$mock->expects( $this->exactly( 3 ) )
+			->method( 'each_callback' )
+			->withConsecutive(
+				[ 1, 'a' ],
+				[ 2, 'b' ],
+				[ 3, 'c' ]
+			);
+
+		// Act
+		$collection->each( [ $mock, 'each_callback' ] );
```
Comment (matipojo) [inline tests/phpunit/elementor/core/utils/test-collection.php:432]: Please create a change on each array item, and assert that the returned collection is updated with these changes.

File: tests/phpunit/elementor/includes/managers/test-image.php [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+class Test_Image extends Elementor_Test_Base {
+
+	public function create_image() {
+		$attachment_id = $this->_make_attachment( [
+			'file' => __DIR__ . '/mock/mock-image.png',
```
Comment (matipojo) [inline tests/phpunit/elementor/includes/managers/test-image.php:15]: Please move the mock image to `/tests/phpunit/resources`

**Review by matipojo — CHANGES_REQUESTED** [review]

File: tests/phpunit/elementor/core/utils/test-collection.php (hunk 11 above)
Comment (StyleShit) [reply tests/phpunit/elementor/core/utils/test-collection.php:432]: The collection shouldn't be updated. In order to update you should use map()

File: core/utils/collection.php (hunk 10 above)
Comment (StyleShit) [reply core/utils/collection.php:130]: I think we need to return $this.
@nahmana check what they do on Laravel

Comment (matipojo) [reply core/utils/collection.php:130]: If it's not changed, it should return `$this`

Comment (matipojo) [reply core/utils/collection.php:130]: @nahmana please add in the description, that the method is not designed to change the array (with a reference to the `map` method)

Comment (nahmana) [reply core/utils/collection.php:130]: Done.

File: tests/phpunit/elementor/core/utils/test-collection.php (hunk 11 above)
Comment (nahmana) [reply tests/phpunit/elementor/core/utils/test-collection.php:432]: Done,

File: tests/phpunit/elementor/includes/managers/test-image.php (hunk 12 above)
Comment (nahmana) [reply tests/phpunit/elementor/includes/managers/test-image.php:15]: Done.
